### PR TITLE
Feature/ecflow suite generator

### PR DESCRIPTION
**Pull request by KyleNevins-NOAA**
**Description**
This PR is to merge in the workflow generator application that uses a YAML file to create suite definition files for ecflow. 

**Type of change**
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

**How Has This Been Tested?**
To test: 
1. Change into the workflow directory: ` cd global-workflow/workflow_generator`
1. Ensure the YAML file is configured with the suites/tasks that are wanted. 
1. Run the utility: `python3 setup_workflow.py --expdir ../parm/config`

** Remaining items **
1. Additional documentation
1. Code documentation
1. Test suite with existing production ecflow definition file.

### CODE REVIEW

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
Looks good in general, except it is lacking in documentation and doesn't appear to follow PEP8. I don't think it is too far off PEP8 and I haven't run a linter yet, but I know there's supposed to be a space after commas, for example.

**Comment by KyleNevins-NOAA** [comment]
> Looks good in general, except it is lacking in documentation and doesn't appear to follow PEP8. I don't think it is too far off PEP8 and I haven't run a linter yet, but I know there's supposed to be a space after commas, for example.

Have made some adjustments to two of the modules but more documentation/pep8/readme updates incoming.

**Comment by KyleNevins-NOAA** [comment]
Added in additional documentation within the code, I believe all methods, classes, and program descriptions are now documented. Also updated the readme.

File: workflow_generator/ecflow_setup/ecflow_definitions.py [hunk 1]
```diff
@@ -117,95 +376,179 @@ def add_repeat(self,repeat,parent=None):
...
+                print(f"The node/trigger {parent}/{trigger} is not available "
+                      f"in suite {self.get_suite_name()}." 
+                      " Please check the configuration file.")
                 print(f"Error {e}")
                 sys.exit(1)
-        print(addTrigger,operand)
         if operand is not None:
-            addTrigger = ecflow.Trigger(addTrigger.get_expression(),operand)
-        self.ecfnodes[parent].add(addTrigger)
+            addTrigger = ecflow.Trigger(addtrigger.get_expression(), operand)
```
Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py]: left side is still using camelcase (should all these be snake case, i.e. add_trigger)?

File: workflow_generator/ecflow_setup/ecflow_definitions.py [hunk 2]
```diff
@@ -423,6 +994,17 @@ def set_max_value(self,range_token):
                 sys.exit(1)
 
     def set_initial_max_value(self,range_token):
```
Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py]: Missing space after comma

Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py]: Missing space after comma

File: workflow_generator/ecflow_setup/ecflow_definitions.py [hunk 3]
```diff
@@ -389,27 +932,55 @@ def __init__(self,ecfItem):
             self.is_list = False
 
     def get_name(self):
+        """
+        Returns the name of the node.
```
Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py:989]: I know it seems redundant/silly, but this and the next few methods should have a Returns section for consistency (especially once we start using a documentation generator to turn these blocks into documentation)

Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py:989]: I know it seems redundant/silly, but this and the next few methods should have a Returns section for consistency (especially once we start using a documentation generator to turn these blocks into documentation)

File: workflow_generator/ecflow_setup/ecflow_definitions.py [hunk 4]
```diff
@@ -771,3 +1759,4 @@ class RangeError(Error):
 class ConfigurationError(Error):
     """Raised when there is an error in the configuration file."""
     pass
+7
```
Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py]: Erroneous 7?

Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py]: Erroneous 7?

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
Descriptions look great. For consistency, I think we should have a Parameters and Returns for every method, even if we just put None in the category.

File: workflow_generator/ecflow_setup/ecflow_definitions.py (hunk 1 above)
Comment (KyleNevins-NOAA) [reply workflow_generator/ecflow_setup/ecflow_definitions.py]: Using snake case now. If you see any others, please let me know and I'll adjust.

File: workflow_generator/ecflow_setup/ecflow_definitions.py (hunk 2 above)
Comment (KyleNevins-NOAA) [reply workflow_generator/ecflow_setup/ecflow_definitions.py]: Shoot, thought I got them all, was bound to miss one or two. Thanks, fixed it up.

File: workflow_generator/ecflow_setup/ecflow_definitions.py (hunk 3 above)
Comment (KyleNevins-NOAA) [reply workflow_generator/ecflow_setup/ecflow_definitions.py:989]: Added in a RETURNS or PARAMETERS section to all of the methods. Added value NONE if there wasn't actually something for that section.

File: workflow_generator/ecflow_setup/ecflow_definitions.py (hunk 4 above)
Comment (KyleNevins-NOAA) [reply workflow_generator/ecflow_setup/ecflow_definitions.py]: Yeah, not sure how that got there. Lucky number 7 has been removed.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**